README.md
# node-dye

[![npm version](https://img.shields.io/npm/v/@archivistnerd/dye.svg)](https://www.npmjs.com/package/@archivistnerd/dye)
[![build status](https://api.travis-ci.com/Archivist-Nerd/node-dye.svg)](https://travis-ci.com/Archivist-Nerd/node-dye.svg)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/Archivist-Nerd/node-dye.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/Archivist-Nerd/node-dye/alerts/)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/Archivist-Nerd/node-dye.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/Archivist-Nerd/node-dye/context:javascript)
[![Known Vulnerabilities](https://snyk.io/test/github/Archivist-Nerd/node-dye/badge.svg)](https://snyk.io/test/github/Archivist-Nerd/node-dye)

Archivist Nerd's dye Package

> Terminal string styling done SIMPLY

## This Package is in VERY EARLY ALPHA
> This package subject to change QUICKLY, and possibly Often

## Installation

```sh
npm install @archivistnerd/dye
```

## Colors
### text colors
* black
* red
* green
* yellow
* blue
* magenta
* cyan
* white
### background colors
* bgblack
* bgred
* bggreen
* bgyellow
* bgblue
* bgmagenta
* bgcyan
* bgwhite

## Usage (Simple)

```js
'use strict';

const dye = require('@archivistnerd/dye');

console.log([
    dye('red','this text is red'),
    dye('bgwhite,blue','blue text, white background')
  ].join('\n')
)
```

## License

MIT


docs/HISTORY.md
# node-testlib
Archivist Nerd's dye



## History
|  Date     | Version  |    type   | Description     |
|-----------|----------|-----------|-----------------|
| 201124.08 |**v0.0.1-alpha**| **removed** | unused require('fs') |
| 201122.04 |**v0.0.1-alpha**| **added** | added bright colors |
| 201121.16 |**v0.0.1-alpha**| **start** | project started |


tests/test.js
'use strict';
const localRequire = true;        //  required from ../src/{package} or ../.
/**
 * @test:package     @archivistnerd/dye
 * @test:name        @archivistnerd/dye/test
 * @test:licence     MIT
 * @test:copyright   Copyright (c) 2020 Archivist-Nerd
 */
let describe = require('@archivistnerd/testlib').describe
  , dye = require( localRequire? '../src/dye': '../.' )
  ;
/**
 * clear the console, and display header
 */
console.clear()
console.log('Testing: @archivistnerd/dye\n')
/**
 * @test:Errors        Test it throws the proper errors
 */
describe( '--all invalid parameters should return empty string', it=>{
  it('no parameters', it=>{
    return ( dye() == '' )
  })

  it('colors not a string', it=>{
    return ( dye(31,'') == '' )
  })

  it('valid colors no text', it=>{
    return ( dye('red') == '' )
  })

  it('valid colors empty text string', it=>{
    return ( dye('red','') == '' )
  })
});
/**
 * @test:Valid        Valid tests, should not throw any errors
 */
describe( '--valid functions', it=>{
  let ansi = (color,code=color) => (color>0)? `\u001B[${code}m`:''
    ;
  it('no valid colors', ()=>{
    let text    = 'is this red?'
      , results = dye(',....,.,.,.,.,,,,....,,.,.,.,.    ,.bmw',text)
      ;
    return ( results == `${text}` )
  })

  it('red text', ()=>{
    let text    = 'is this red?'
      , results = dye('red',text)
      ;
    return ( results == `${ansi(31)}is this red?${ansi(39)}` )
  })

  it('background white', ()=>{
    let text    = 'is this all white?'
      , results = dye('bgwhite',text)
      ;
    return ( results == `${ansi(47)}${text}${ansi(49)}` )
  })

  it('combo: white background, red text', ()=>{
    let text    = 'is this text red on a white background?'
      , results = dye('bgwhite.red',text)
      ;
    return ( results == `${ansi(47)+ansi(31)}${text}${ansi(39)+ansi(49)}` )
  })

  it('combo: many options (keep last valid)', ()=>{
    let text    = 'is this text red on a white background?'
      , results = dye('bgblack.blue.bggreen.yellow.bgwhite.green.yellow.red',text)
      ;
    return ( results == `${ansi(47)+ansi(31)}${text}${ansi(39)+ansi(49)}` )
  })

  it('mixes comma and period seperators', ()=>{
    let text    = 'is this text red on a white background?'
      , results = dye('bgblack,blue.bggreen.yellow,bgwhite.green,yellow.red,,',text)
      ;
    return ( results == `${ansi(47)+ansi(31)}${text}${ansi(39)+ansi(49)}` )
  })
});

// console.log( process.env )

/**
 *   output success/fail results
 */
if ( describe().length===0 )
  console.log('\n\n   *** ALL TESTS SUCCESSFUL ***');
else {
  console.log('\n\n   *** !!!SOME TESTS FAILED!!! ***')
  console.log( describe() )
  process.exit(1)
}

src/dye.js
'use strict';
/**
 * @api             dye
 * @api:npm         @archivistnerd/dye
 * @api:git         https://github.com/Archivist-Nerd/node-dye
 * @api:Licence     MIT
 * @api:Copyright   Copyright (c) 2020 Archivist-Nerd
 *
 *
 * @api:Example:
 *      const dye = require('@archivistnerd/dye');
 */
/**
 * @api:name    dye    returns string with ansi color codes
 * @api:group   functions
 *
 * @api:Param   {String} | {boolean}    string: color options
 *                                      boolean: true=force colors, false=ignore colors
 * @api:Param   {String}                text to enclose in above color options
 *
 * @api:Example:
 *      dye('green','It Works :)')
 */
const colors = {
        //            text colors               end code: 39
        black:   30,
        red:     31,
        green:   32,
        yellow:  33,
        blue:    34,
        magenta: 35,
        cyan:    36,
        white:   37,
        //            text bright colors        end code: 39
        grey:          90,
        gray:          90,
        blackBright:   90,
        redBright:     91,
        greenBright:   92,
        yellowBright:  93,
        blueBright:    94,
        magentaBright: 95,
        cyanBright:    96,
        whiteBright:   97,
        //            background colors         end code: 49
        bgblack:   40,
        bgred:     41,
        bggreen:   42,
        bgyellow:  43,
        bgblue:    44,
        bgmagenta: 45,
        bgcyan:    46,
        bgwhite:   47,
        //            background bright colors  end code: 49
        bggrey:          100,
        bggray:          100,
        bgblackBright:   100,
        bgredBright:     101,
        bggreenBright:   102,
        bgyellowBright:  103,
        bgblueBright:    104,
        bgmagentaBright: 105,
        bgcyanBright:    106,
        bgwhiteBright:   107,
      }
    , ansi = (color,code=color) => (color>0)? `\u001B[${code}m`:''
    , forceColors = true
    ;
module.exports = function dye( choices='', text='' ){
  if ( choices === true || choices === false) return ignoreColors = choices
  if ( !forceColors || typeof choices!=='string' || choices=='' || text=='') return text

  let fg=0
    , bg=0
    ;

  choices.replace(/,/g,'.').split('.').forEach( color => {
    let code = colors[ color.toLowerCase() ]
      , group = (code/10>>0)        // returns only integer of division
      ;
    if ( code == undefined) return
    if ( group == 3 || group == 9)
      fg = code; 
    else
      bg = code;
  })

  return `${ansi(bg)+ansi(fg)}${text}${ansi(fg,39)+ansi(bg,49)}`
}